src/ring/adapter/jetty9/common.clj

```diff
@@ -1,36 +1,36 @@
 (ns ring.adapter.jetty9.common
-  (:require [clojure.string :as string])
-  (:import [jakarta.servlet.http HttpServletRequest HttpServletResponse]
+  (:require [clojure.string :as string]
+            [ring.core.protocols :as protocols])
+  (:import [org.eclipse.jetty.http HttpHeader HttpField MimeTypes]
+           [org.eclipse.jetty.io.content
+            ContentSourceInputStream ContentSinkOutputStream]
+           [org.eclipse.jetty.server Request Response]
+
            [java.util Locale]))
 
 (defprotocol RequestMapDecoder
   (build-request-map [r]))
 
-(defn set-headers
-  "Update a HttpServletResponse with a map of headers."
-  [^HttpServletResponse response, headers]
-  (doseq [[key val-or-vals] headers]
-    (if (string? val-or-vals)
-      (.setHeader response key val-or-vals)
-      (doseq [val val-or-vals]
-        (.addHeader response key val))))
-  ; Some headers must be set through specific methods
-  (some->> (get headers "Content-Type")
-           (.setContentType response)))
+(defn set-headers!
+  "Update response with a map of headers."
+  [^Response response headers]
+  (let [header-writer (.getHeaders response)]
```

Comment:
Type hint to avoid reflection, maybe ^HttpFields$Mutable ... or there might be an interface that's a better choice

Reply:
Just verified there is no reflection with this header-writer